File: Exercise_5.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <stack> 
using namespace std; 
  
// A utility function to swap two elements 
void swap(int* a, int* b) 
{ 
    int t = *a; 
    *a = *b; 
    *b = t; 
} 
  
/* This function is same in both iterative and recursive*/
int partition(int arr[], int l, int h) 
{ 
    //Do the comparison and swapping here 
    int pivot = arr[h];
    int i = l - 1;

    for(int j = l; j < h; j++) {
        if(arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[h]);
    return i + 1;
} 
  
/* A[] --> Array to be sorted,  
l --> Starting index,  
h --> Ending index */
void quickSortIterative(int arr[], int l, int h) 
{ 
    //Try to think that how you can use stack here to remove recursion.
    stack<pair<int, int>> s;
    s.push({l, h});
    while(!s.empty()) {
        auto range = s.top();
        s.pop();
        l = range.first;
        h = range.second;
        int p = partition(arr, l, h);
        if(p - 1 > l)
            s.push({l, p - 1});
        if(p + 1 < h)
            s.push({p + 1, h});
    }
} 
  
// A utility function to print contents of arr 
void printArr(int arr[], int n) 
{ 
    int i; 
    for (i = 0; i < n; ++i) 
        cout << arr[i] << " "; 
} 
  
// Driver code 
int main() 
{ 
    int arr[] = { 4, 3, 5, 2, 1, 3, 2, 3 }; 
    int n = sizeof(arr) / sizeof(*arr); 
    quickSortIterative(arr, 0, n - 1); 
    printArr(arr, n); 
    return 0; 
} 

File: Exercise_5.py
# Python program for implementation of Quicksort

# This function is same in both iterative and recursive
def partition(arr, l, h):
  #write your code here
  pivot = arr[h]
  i = l - 1

  for j in range(l, h):
    if arr[j] <= pivot:
      i += 1
      arr[i], arr[j] = arr[j], arr[i]
  
  arr[i + 1], arr[h] = arr[h], arr[i + 1]
  return i + 1


def quickSortIterative(arr, l, h):
  #write your code here
  stack = []
  stack.append((l, h))

  while stack:
    l, h = stack.pop()
    p = partition(arr, l, h)
    
    if p - 1 > l:
      stack.append((l, p -1))

    if p+ 1 < h:
      stack.append((p + 1, h))

if __name__ == "__main__":
  arr= [4, 3, 5, 2, 1, 3, 2, 3]
  n = len(arr)
  print("Given array is: ")
  print(arr)
  quickSortIterative(arr, 0, n - 1)
  print("Sorted array is: ")
  print(arr)